### PR TITLE
php: Fix CVE-2018-17082 (nixos-unstable)

### PR DESCRIPTION
**Pull request by Ekleog**
#### php: 7.2.10 -> 7.2.12, 7.1.22 -> 7.1.24

Also make Darwin align itself to Linux versions, due to CVE-2018-17082
forcing our hand on this. This means Darwin must compile without intl.

###### Motivation for this change

The alternative would be to backport the fix to a previous version of `php`. It looks [pretty easy](https://github.com/php/php-src/commit/23b057742e3cf199612fa8050ae86cae675e214e), but I know the PHP code base very little and wouldn't trust simply using this patch to fix the issue unless spending much more time on it.

cc @etu who introduced the split between darwin and linux

@GrahamcOfBorg build php71 php72
@GrahamcOfBorg test elk owncloud

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Ekleog** [comment]
See also backports at https://github.com/NixOS/nixpkgs/pull/50505 and (almost-a-backport) https://github.com/NixOS/nixpkgs/pull/50377.

**Comment by etu** [comment]
@Ekleog I first used a non-assert solution and got it pointed out to me that that was a bad idea because then it goes unnoted that the change of feature flags have happened. Which I agree to.

That's why it ended up like this but got documented. And I've been following the issue on the PHP bugtracker to see if they actually will resolve this but it's not very hopeful. Which actually kinda surprise me since they broke the build in a minor release and don't seem able to fix it.

I'm also in favor of aligning it up again and if you and others think that this assert warning solution is good enough for this, I'm all for it.

I've also looked through the intl extension in the docs a bit https://secure.php.net/intl and, I must say. From my years (8+) of working professionally with PHP. I don't think that I've ever to my knowledge have used this extension myself. It may be used by some popular libraries, I don't know about that.

But overall I'm in favor of this.

With this fix we could close #50014.

**Comment by Ekleog** [comment]
@etu Hmm, it's possible to add the `assert` and require that `intl` is manually set to `false`… would that do the job? It may be a harsh bump for stable-channel users, but it's likely better than leaving it go unnoticed indeed? Also, Darwin users are more or less pushed towards `nixpkgs-unstable` anyway IIRC, so it may be less big a deal than it'd have been had it been on all linuxes.

**Comment by etu** [comment]
@Ekleog As the person who pointed me out to me said: "Better to fail on rebuild than the user upgrading and go home for the weekend and see it fail while it's weekend."

**Comment by etu** [comment]
@Ekleog Status on this?

**Comment by Ekleog** [comment]
@etu Sorry for the delay, I was away from the Internet last week! I think unstable should go without the manual change, because the default will need to change before 19.03 anyway, and `unstable` users should be aware of this kind of things.

So I've pushed two new versions of this PR, for `unstable` with the change of default value and an error if `intl` is forced to true, and for `18.09` with no change of the default value and an error if `intl` is not forced to false.

This will likely break tests for 18.09, unfortunately… but I don't really see a better way forward, as silent breakage or leaving a security vulnerability are two important issues… I guess we'll have to fix tests once we notice they break, though I'm not sure how often tests are run for Darwin on 18.09.

TL;DR: I think this is good to go, and we'll have to handle the broken stuff when it becomes apparent.

**Comment by etu** [comment]
Yeah, there's those two choices. And neither choice is a good choice.

So one have to pick what's best. And it this case I agree that it's different on unstable and stable :slightly_smiling_face:

**Review by etu — APPROVED** [review]

**Comment by delroth** [comment]
Counterproposal on #51091 to keep backwards compatibility -- the patches are trivial enough that imo it's worth it. I'll let you pick your favorite option :)

**Comment by Ekleog** [comment]
Superseded by now-merged #51091
